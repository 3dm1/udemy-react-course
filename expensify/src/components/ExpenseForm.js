import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    note: this.props.expense ? this.props.expense.note : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused : focused }));
  }

  onFormSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description) {
      this.setState(() => ({ error: 'Please input a description.' }));
    } else if (!this.state.amount) {
      this.setState(() => ({ error : 'Please input an amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onFormSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={(day) => false}
          block
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
          >
        </textarea>
        <div>
          <button className="button">{this.props.expense ? 'Edit Expense' : 'Add Expense'}</button>
        </div>
      </form>
    );
  }
};
