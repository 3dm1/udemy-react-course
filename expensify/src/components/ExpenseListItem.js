import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({ expense }) => (
  <div>
    <Link to={`/edit/${expense.id}`}>
      <h3>{expense.description}</h3>
    </Link>
    <p>{expense.amount} - {expense.createdAt}</p>
  </div>
);

export default ExpenseListItem;
