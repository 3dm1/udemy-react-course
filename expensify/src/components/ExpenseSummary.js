import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpenseSummary extends React.Component {

  render() {
    const formattedTotal = numeral(this.props.expenseTotal / 100).format('$0,0.00');
    let message;
    switch (this.props.expenseCount) {
      case 0:
        message = 'No expenses found';
        break;
      case 1:
        message = `Viewing ${this.props.expenseCount} expense totalling ${formattedTotal}`;
        break;
      default:
        message = `Viewing ${this.props.expenseCount} expenses totalling ${formattedTotal}`;
    }

    return (
      <div>
        <h1>{message}</h1>
      </div>
    )
  }
};

const mapStateToProps = ({ expenses, filters }) => {
  const visibleExpenses = getVisibleExpenses(expenses, filters);
  return {
    expenseCount : visibleExpenses.length,
    expenseTotal : selectExpensesTotal(visibleExpenses)
  }
}


export default connect(mapStateToProps)(ExpenseSummary)
