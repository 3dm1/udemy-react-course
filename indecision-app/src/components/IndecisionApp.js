import React from 'react';
import Header from './Header';
import Action from './Action';
import Options from './Options';
import AddOption from './AddOption';
import OptionModal from './OptionModal'

export default class IndecisionApp extends React.Component {
  state = {
    options : [],
    selectedOption : undefined
  };

  componentDidMount() {
    try {
      const options = JSON.parse(localStorage.getItem('options'));
      if (options) {
        this.setState(() => ({ options }));
      }
    } catch (e) {
      console.log(e);
      //do nothing
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.options.length != this.state.options.length) {
      localStorage.setItem('options', JSON.stringify(this.state.options));
    }
  }

  handleDeleteOption = (option) => {
    this.setState((prevState) => ({
      options: prevState.options.filter((item) => item !== option)
    }))
  };

  handleDeleteOptions = () => {
    this.setState(() => ({ options: [] }));
  };

  handleAction = () => {
    const options = this.state.options;
    const selection = Math.floor(Math.random() * options.length);
    this.setState(() => ({
      selectedOption : options[selection]
    }));
  }

  handleAddOption = (option) => {
    if (!option) {
      return 'Enter valid value to add item';
    } else if(this.state.options.indexOf(option) > -1) {
      return 'This option already exists';
    }

    this.setState((prevState) => ({ options: [...prevState.options, option] }))
  };

  handleConfirmOption = () => {
    this.setState(() => ({ selectedOption : undefined }));
  }

  render() {
    const subtitle = 'Put your life in the hands of a computer';

    return (
      <div>
        <Header subtitle={subtitle}/>
        <div className="container">
          <Action
            hasOptions={this.state.options.length > 0}
            handleAction={this.handleAction}
          />
          <div className="widget">
            <Options
              options={this.state.options}
              handleDeleteOption={this.handleDeleteOption}
              handleDeleteOptions={this.handleDeleteOptions}
            />
            <AddOption handleAddOption={this.handleAddOption}/>
          </div>
        </div>
        <OptionModal
          selectedOption={this.state.selectedOption}
          handleConfirmOption={this.handleConfirmOption}
        />
      </div>
    );
  }
}
