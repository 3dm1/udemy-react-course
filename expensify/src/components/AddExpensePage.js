import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { createExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  createExpense = (expense) => {
    this.props.createExpense(expense)
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        This is from the expense component
        <ExpenseForm onSubmit= {this.createExpense} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  createExpense: (expense) => dispatch(createExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
