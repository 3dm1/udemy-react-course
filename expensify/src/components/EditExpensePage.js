import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  editExpense = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  removeExpense = () => {
    this.props.removeExpense({ id : this.props.expense.id });
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit= {(expense) => this.editExpense(expense)}
        />
        <button
          onClick={this.removeExpense}
          >
            Remove
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (expense) => dispatch(removeExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
