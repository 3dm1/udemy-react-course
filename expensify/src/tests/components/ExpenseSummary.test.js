import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render correctly with no visible expenses', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expenseCount={0}
      expenseTotal={0}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should render correctly with a single expense', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expenseCount={1}
      expenseTotal={1295}
    />);
  expect(wrapper).toMatchSnapshot();
});

test('should render correctly with multiple expenses', () => {
  const wrapper = shallow(
    <ExpenseSummary
      expenseCount={7}
      expenseTotal={9425}
    />);
  expect(wrapper).toMatchSnapshot();
});
