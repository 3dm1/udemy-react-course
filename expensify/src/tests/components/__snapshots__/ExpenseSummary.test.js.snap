// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with a single expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <div>
        Viewing 
        <span>
          1
        </span>
         
        expense
         totalling 
        <span>
          $12.95
        </span>
      </div>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <div>
        Viewing 
        <span>
          7
        </span>
         
        expenses
         totalling 
        <span>
          $94.25
        </span>
      </div>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render correctly with no visible expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <div>
        No expenses found
      </div>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
